### PR TITLE
permit opt tokenizer

### PR DESCRIPTION
**Pull request by bmosaicml**
# What does this PR do?
Add functionality to support ICL with OPT tokenizer (which prepends special tokens each time tokenizer is called). Due to the way we concatenate context and continuation encodings into a single input, it's important that no special tokens are put between them.

GPT Neo 1.3 B gets following accuracy
metrics/piqa/5-shot/InContextLearningMultipleChoiceAccuracy: 0.717934787273407
metrics/lambada/0-shot/InContextLearningLMAccuracy: 0.5883721113204956

OPT 1.3B gets:
metrics/piqa/5-shot/InContextLearningMultipleChoiceAccuracy: 0.719565212726593
metrics/lambada/0-shot/InContextLearningLMAccuracy: 0.5883721113204956

MosaicGPT 1.3B gets:
metrics/piqa/5-shot/InContextLearningMultipleChoiceAccuracy: 0.637499988079071
metrics/lambada/0-shot/InContextLearningLMAccuracy: 0.4075581431388855

# What issue(s) does this change relate to?
 ICL eval doesn't work with OPT tokenizer since it adds special tokens between contexts/continuations.

# Before submitting
- [ x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ x] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x ] Did you update any related docs and document your change?
- [x ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ x] Did you run the tests locally to make sure they pass?
- [ x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/datasets/in_context_learning_evaluation.py [hunk 1]
```diff
@@ -153,10 +153,10 @@ def prep_examples(self, num_fewshot: int, prompt_string: str, example_delimiter:
 
             cont = f'{continuation_delimiter}{cont}'
 
-            encoded_example['context'] = self.tokenizer(ctxt)
-            encoded_example['continuation'] = self.tokenizer(cont)
             encoded_example['preamble'] = self.tokenizer(
                 preamble)  # if the preamble is empty then these will be 0-length lists
+            encoded_example['context'] = self.tokenizer(ctxt, add_special_tokens=len(preamble) == 0)
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:158]: I believe that when you tokenize an empty string, it still adds special tokens
```In [4]: tt = transformers.AutoTokenizer.from_pretrained('facebook/opt-125m')

In [5]: tt('')
Out[5]: {'input_ids': [2], 'attention_mask': [1]}```

So, your code still gets the right outcome I think, but it would be clearer to write it such that the `encoded_example['preamble']` is just an empty list if `preamble` is empty.

File: composer/datasets/in_context_learning_evaluation.py [hunk 2]
```diff
@@ -300,11 +300,11 @@ def prep_examples(self, num_fewshot: int, prompt_string: str, example_delimiter:
                 query = f'{example_delimiter}{query}'
 
             choices = [f'{continuation_delimiter}{choice}' for choice in choices]
-            encoded_example['query'] = self.tokenizer(query)
-            encoded_example['choices'] = [self.tokenizer(choice) for choice in choices]
             encoded_example['preamble'] = self.tokenizer(
                 preamble)  # if the preamble is empty then these will be 0-length lists
             encoded_example['gold_idx'] = gold_idx
+            encoded_example['query'] = self.tokenizer(query, add_special_tokens=len(preamble) == 0)
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:306]: same comment as above.

**Review by dakinggg — APPROVED** [review]
LGTM, see the one comment

File: composer/datasets/in_context_learning_evaluation.py (hunk 2 above)
Comment (bmosaicml) [reply composer/datasets/in_context_learning_evaluation.py:306]: Good catch, I was actually prepending two <s/> tokens in the case of zero-shot eval, so it is fixed now!

File: composer/datasets/in_context_learning_evaluation.py [hunk 3]
```diff
@@ -153,10 +153,10 @@ def prep_examples(self, num_fewshot: int, prompt_string: str, example_delimiter:
 
             cont = f'{continuation_delimiter}{cont}'
 
-            encoded_example['context'] = self.tokenizer(ctxt)
-            encoded_example['continuation'] = self.tokenizer(cont)
             encoded_example['preamble'] = self.tokenizer(
                 preamble)  # if the preamble is empty then these will be 0-length lists
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:157]: could you correct the comment too? to note that it may contain just a special token

File: composer/datasets/in_context_learning_evaluation.py [hunk 4]
```diff
@@ -298,13 +298,12 @@ def prep_examples(self, num_fewshot: int, prompt_string: str, example_delimiter:
                 'choices'], self.samples[sample_idx]['gold'],
             if len(preamble) > 0:
                 query = f'{example_delimiter}{query}'
-
             choices = [f'{continuation_delimiter}{choice}' for choice in choices]
-            encoded_example['query'] = self.tokenizer(query)
-            encoded_example['choices'] = [self.tokenizer(choice) for choice in choices]
             encoded_example['preamble'] = self.tokenizer(
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:302]: same

File: tests/datasets/test_in_context_learning_datasets.py [hunk 5]
```diff
@@ -45,6 +46,41 @@ def test_lm_task_dataloader(dataset_uri, tiny_gpt2_tokenizer):
...
+    assert 'continuation_indices' in batch
+    assert isinstance(batch['continuation_indices'], list) and len(batch['continuation_indices']) == batch_size
+    assert 'mode' in batch
+    assert batch['mode'] == 'icl_task'
+    min_idx = min(batch['continuation_indices'][0]).item()
+    max_idx = max(batch['continuation_indices'][0]).item()
+    assert tokenizer.decode(batch['input_ids'][0][min_idx:max_idx + 1]) == ' glen'
+
+
+@pytest.mark.parametrize('dataset_uri', ['lambada_small.jsonl'])
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:76]: could you add a zero shot test that would've caught the previous issue?

Comment (dakinggg) [reply tests/datasets/test_in_context_learning_datasets.py:76]: oh i see you changed these to zero shot. in that case, could you add a test for >0 shots too?

File: composer/datasets/in_context_learning_evaluation.py (hunk 2 above)
Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py]: why was it ending up with doubled special tokens? just for my undestanding. it looks like you would have one at the start of preamble (whether preamble was empty or not), and then where would the other one come from?

**Comment by abhi-mosaic** [comment]
@bmosaicml aren't the OPT scores a bit low? From our internal data, I see:

piqa 0-shot (not 5-shot): 0.7236
lambada 0-shot: 0.588


Vs:

> 
> OPT 1.3B gets:
> metrics/piqa/5-shot/InContextLearningMultipleChoiceAccuracy: 0.626630425453186
> metrics/lambada/0-shot/InContextLearningLMAccuracy: 0.37538760900497437

**Comment by bmosaicml** [comment]
> @bmosaicml aren't the OPT scores a bit low? From our internal data, I see:
> 
> OPT 1.3B piqa 0-shot (not 5-shot): 0.7236 OPT 1.3B lambada 0-shot: 0.588
> 
> Vs:
> 
> > OPT 1.3B gets:
> > metrics/piqa/5-shot/InContextLearningMultipleChoiceAccuracy: 0.626630425453186
> > metrics/lambada/0-shot/InContextLearningLMAccuracy: 0.37538760900497437

yes they are a bit low....where are you getting your numbers from?

**Comment by bmosaicml** [comment]
@dakinggg  @abhi-mosaic  Sorry, the initial numbers I wrote were for OPT 125m I just retested with OPT 1.3B and the numbers match what we expect!

**Review by dakinggg — APPROVED** [review]
Fantastic! re-approving :)

**Review by karan6181 — APPROVED** [review]
